File: uhrwerk-config/src/main/java/io/qimia/uhrwerk/config/representation/Dag.java
package io.qimia.uhrwerk.config.representation;

import java.util.Arrays;
import io.qimia.uhrwerk.config.ConfigException;

public class Dag {
  private Connection[] connections;
  private Table[] tables;

  public Connection[] getConnections() {
    return connections;
  }

  public void setConnections(Connection[] connections) {
    this.connections = connections;
  }

  public Table[] getTables() {
    return tables;
  }

  public void setTables(Table[] tables) {
    this.tables = tables;
  }

  public void validate(String path) {
    path += "/";
    if(connections==null){
      throw new ConfigException("Missing field:" + path + "connections");
    }
    else{
      for (Connection c: connections){
        c.validate(path);
      }
    }
    if(tables==null){
      throw new ConfigException("Missing field:" + path + "tables");
    }
    else{
      for (Table t: tables){
        t.validate(path);
      }
    }
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Dag dag = (Dag) o;
    return Arrays.equals(connections, dag.connections) &&
            Arrays.equals(tables, dag.tables);
  }

  @Override
  public int hashCode() {
    int result = Arrays.hashCode(connections);
    result = 31 * result + Arrays.hashCode(tables);
    return result;
  }

  @Override
  public String toString() {
    return "Dag{" +
            "connections=" + Arrays.toString(connections) +
            ", tables=" + Arrays.toString(tables) +
            '}';
  }
}

File: uhrwerk-config/src/main/kotlin/io/qimia/uhrwerk/config/representation/Dag.kt
package io.qimia.uhrwerk.config.representation

import io.qimia.uhrwerk.config.builders.ConfigException

data class Dag(
    var secrets: Array<Secret>? = null,
    var connections: Array<Connection>? = null,
    var tables: Array<Table>? = null
) {
    fun validate(path: String) {
        var path = path
        path += "/"
        if (connections == null) {
            throw ConfigException("Missing field:" + path + "connections")
        } else {
            for (c in connections!!) {
                c.validate(path)
            }
        }
        if (tables == null) {
            throw ConfigException("Missing field:" + path + "tables")
        } else {
            for (t in tables!!) {
                t.validate(path)
            }
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Dag) return false

        if (secrets != null) {
            if (other.secrets == null) return false
            if (!secrets.contentEquals(other.secrets)) return false
        } else if (other.secrets != null) return false
        if (connections != null) {
            if (other.connections == null) return false
            if (!connections.contentEquals(other.connections)) return false
        } else if (other.connections != null) return false
        if (tables != null) {
            if (other.tables == null) return false
            if (!tables.contentEquals(other.tables)) return false
        } else if (other.tables != null) return false

        return true
    }

    override fun hashCode(): Int {
        var result = secrets?.contentHashCode() ?: 0
        result = 31 * result + (connections?.contentHashCode() ?: 0)
        result = 31 * result + (tables?.contentHashCode() ?: 0)
        return result
    }

    override fun toString(): String {
        return "Dag(secrets=${secrets?.contentToString()}, connections=${connections?.contentToString()}, tables=${tables?.contentToString()})"
    }
}